core/templates/dev/head/pages/exploration_player/AudioControlsDirective.js

```diff
@@ -56,20 +59,37 @@ oppia.directive('audioControls', [
               '/icons/rewind-five.svg'));
 
           $scope.isAudioAvailableInCurrentLanguage = function() {
-            return Boolean(getCurrentAudioTranslation());
+            return Boolean(getAudioTranslationInCurrentLanguage());
           };
 
           $scope.doesCurrentAudioTranslationNeedUpdate = function() {
-            return getCurrentAudioTranslation().needsUpdate;
+            return getAudioTranslationInCurrentLanguage().needsUpdate;
           };
 
-          $scope.playPauseAudioTranslation = function() {
-            // TODO(tjiang11): Change from on-demand loading to pre-loading.
+          $scope.onSpeakerIconClicked = function() {
+            var audioTranslation = getAudioTranslationInCurrentLanguage();
+            if (audioTranslation) {
+              if (!AudioPreloaderService.hasPreloadedLanguage(
+                    getCurrentAudioLanguageCode()) &&
+                    !isCached(audioTranslation)) {
+                AudioPreloaderService.excludeFile(
```

Comment:
Mmm ... I don't really see why the audio controls directive cares about all these details. Shouldn't the preloader (or even the backend api service) handle it internally? I assume the issue is that the current translation is in the middle of being loaded, but presumably the backend api service could maintain some sort of list detailing which requests are currently in flight?